sources/core/Xenko.Core/Threading/ThreadPool.cs

```diff
@@ -17,93 +17,126 @@ namespace Xenko.Core.Threading
     {
-        private const long MaxIdleTimeInTicks = 5 * TimeSpan.TicksPerSecond;
+        private const int MaxIdleTimeInMS = 5000;
+        private readonly long MaxIdleTimeTS = (long)((double)Stopwatch.Frequency / 1000 * MaxIdleTimeInMS);
 
         public static readonly ThreadPool Instance = new ThreadPool();
 
+        private readonly Action<object> cachedTaskLoop;
+
         private readonly int maxThreadCount = Environment.ProcessorCount + 2;
         private readonly Queue<Action> workItems = new Queue<Action>();
         private readonly ManualResetEvent workAvailable = new ManualResetEvent(false);
 
         private SpinLock spinLock = new SpinLock();
-        private int workingCount;
-        /// <summary> Usage only within <see cref="spinLock"/> </summary>
+        private int busyCount;
         private int aliveCount;
 
+        public ThreadPool()
+        {
+            // Cache delegate to avoid pointless allocation
+            cachedTaskLoop = (o) => ProcessWorkItems();
+        }
+
         public void QueueWorkItem([NotNull] [Pooled] Action workItem)
         {
-            var lockTaken = false;
+            bool lockTaken = false;
+            bool startNewTask = false;
+            PooledDelegateHelper.AddReference(workItem);
             try
             {
                 spinLock.Enter(ref lockTaken);
-
-                PooledDelegateHelper.AddReference(workItem);
                 workItems.Enqueue(workItem);
                 workAvailable.Set();
 
-                var curWorkingCount = Interlocked.CompareExchange(ref workingCount, 0, 0);
-                if (curWorkingCount + 1 >= aliveCount && aliveCount < maxThreadCount)
+                // We're only locking when potentially increasing aliveCount as we
+                // don't want to go above our maximum amount of threads.
+                int curBusyCount = Interlocked.CompareExchange(ref busyCount, 0, 0);
+                int curAliveCount = Interlocked.CompareExchange(ref aliveCount, 0, 0);
+                if (curBusyCount + 1 >= curAliveCount && curAliveCount < maxThreadCount)
                 {
-                    aliveCount++;
-                    Task.Factory.StartNew(ProcessWorkItems, TaskCreationOptions.LongRunning);
+                    // Start threads as busy otherwise only one thread will be created 
+                    // when calling this function multiple times in a row
+                    Interlocked.Increment(ref busyCount);
+                    Interlocked.Increment(ref aliveCount);
+                    startNewTask = true;
                 }
             }
             finally
             {
                 if (lockTaken)
+                {
                     spinLock.Exit(true);
+                }
+            }
+            // No point in wasting spins on the lock while creating the task
+            if (startNewTask)
+            {
+                new Task(cachedTaskLoop, null, TaskCreationOptions.LongRunning).Start();
```

Comment:
Why that change? It is probably even worst than using TaskFactory.StartNew().

A long running task is a task that is blocking (e.g. using a Thread.Join() or Thread.Sleep()). That is not the case here. Over subscribing the managed thread pool is a recipe for disaster.

Reply:
That task *is* blocking, either through ``workAvailable.WaitOne`` or the spinwait. That task lives for at least 5 seconds and in most case for the lifetime of the application, it is long running.
What we're doing here is effectively the same thing as starting a thread when we have too much work to deal with for the current amount of thread and let them block while waiting for work.

Reply:
OK for that remark. I should have read more carefully.

Still, the `new Task().Start()` needs to go away as it introduces unnecessary locking (as per the link I provided).